L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,450 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): ZhenLian
+* Approver: markdroth, yashykt, jiangtaoli2016
+* Status: Draft
+* Implemented in:
+* Last updated: September 9, 2020
+
+## Abstract
+
+This proposal aims to provide the following security features in a new TLS credential API:
+1. credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down
+2. custom authorization check: optionally bypass the certificate check, bypass the hostname check on the client side, and perform an additional per-connection authorization callback at the end of authentication
+3. TLS version configuration: optionally configure the minimum and maximum of the TLS version that will be used in the handshake
+
+## Background
+
+The current TLS implementation in gRPC C core has some restrictions:
+1. lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down
+2. inflexibility in the peer verification. In a mutual TLS scenario, servers would always perform certificate verification, and clients would always perform certificate verification as well as default hostname check. No customized checks can be introduced.
+3. inability to choose the TLS version that will be used
```

Comment:
Maybe rephrase it to "inability to choose the maximum TLS version supported in OpenSSL/boringSSL for use""? as I believe users always prefer TLS 1.3 over TLS 1.2 when both are available?

Reply:
Sure, I added "maximum/minimum" instead of "maximum" though since that's the API implies...

Reply:
I think the purpose of specifying max and min TLS is a little different. For max, it should be set to the maximum version a client advocates (e.g., TLS 1.3) so that if both sides support the same maximum, the maximum will be selected and used. For min, it is the bottom line a user can tolerate and if the negotiated version is smaller than it, the TLS handshake should fail. @jiangtaoli2016 for confirm. 
How about saying "inability to set the minimum and maximum supported protocol versions for SSL CTX"?  Sorry for the pickiness, but I just want to make it clear.

Reply:
No all customers prefer TLS 1.3. The purpose of setting max/min TLS version is to give users some control on TLS version, e.g., if user only wants TLS 1.3 or only wants TLS 1.2. Of course, such API is not guarantee to be working. If the underlying OpenSSL library uses 1.0.2, set min/max TLS version won't take effect.

Reply:
Understood. So I will take it as a confirmation on my understanding.